core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -339,64 +350,77 @@ class Partition(val topicPartition: TopicPartition,
     }
   }
 
-  private def getLocalReplica(replicaId: Int,
-                              currentLeaderEpoch: Optional[Integer],
-                              requireLeader: Boolean): Either[Replica, Errors] = {
+  private def getLocalLog(currentLeaderEpoch: Optional[Integer],
+                         requireLeader: Boolean): Either[Log, Errors] = {
     checkCurrentLeaderEpoch(currentLeaderEpoch) match {
       case Errors.NONE =>
-        if (requireLeader && !leaderReplicaIdOpt.contains(localBrokerId)) {
+        if (requireLeader && !isLeader) {
           Right(Errors.NOT_LEADER_FOR_PARTITION)
         } else {
-          val replica = allReplicasMap.get(replicaId)
-          if (replica == null) {
-            if (requireLeader)
-              Right(Errors.NOT_LEADER_FOR_PARTITION)
-            else
-              Right(Errors.REPLICA_NOT_AVAILABLE)
-          } else {
-            Left(replica)
+          log match {
+            case Some(leaderLog) =>
+              Left(leaderLog)
+            case _ =>
+              if (requireLeader)
+                Right(Errors.NOT_LEADER_FOR_PARTITION)
+              else
+                Right(Errors.REPLICA_NOT_AVAILABLE)
           }
         }
       case error =>
         Right(error)
     }
   }
 
-  def localReplica: Option[Replica] = getReplica(localBrokerId)
-
-  def localReplicaOrException: Replica = localReplica.getOrElse {
-    throw new ReplicaNotAvailableException(s"Replica for partition $topicPartition is not available " +
+  def localLogOrException: Log = log.getOrElse {
+    throw new ReplicaNotAvailableException(s"Log for partition $topicPartition is not available " +
       s"on broker $localBrokerId")
   }
 
-  def futureLocalReplica: Option[Replica] = getReplica(Request.FutureLocalReplicaId)
+  private def localReplica: Option[LocalReplica] = getReplica(localBrokerId).asInstanceOf[Option[LocalReplica]]
```

Comment:
I wonder if we should model this by having a `LocalReplica` as a private field of `Partition` and changing `allReplicaMap` to only contain remote replicas (changing the name of the variable in the process).

Reply:
I thought about this too, but decided to keep the original behavior (just to keep scope of change small). If we want to move local replica out of collection, we need to analyze code to see if clients need to make this distinction and how that impacts all code dealing with replica. I don't know that yet. Let me know how you feel and I will do the analysis.

Reply:
So I tried to separate it out and its reasonably large change. As current change is big enough, I think if we want this, it should be done as a separate change. Let me know what you think.

Reply:
> In fact, I think it might be possible to turn `Replica` into a trait and then let `Log` implement it directly. Then we could get rid of `LocalReplica`. That would also help us clean up `RemoteReplica`, since the usage of `LogOffsetMetadata` only makes sense for the local replica.

This suggestion looks promising.

Reply:
Filed https://issues.apache.org/jira/browse/KAFKA-8525 to track this work as a separate task.